src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_CodeLens.cs

```diff
@@ -63,12 +87,140 @@ public Task<string> GetFullyQualifiedName(DocumentId documentId, TextSpan textSp
+            private readonly DocumentId _documentId;
+
+            private VersionStamp _lastVersion;
+            private ResettableDelay _resettableDelay;
+
+            public static async Task TrackAsync(CodeAnalysisService owner, Workspace workspace, DocumentId documentId, CancellationToken cancellationToken)
+            {
+                var document = workspace.CurrentSolution.GetDocument(documentId);
+                if (document == null)
+                {
+                    return;
+                }
+
+                // if anything under the project this file belong to changes, then invalidate the code lens so that it can refresh
+                var dependentVersion = await document.Project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false);
+                var _ = new WorkspaceChangeTracker(owner, workspace, documentId, dependentVersion);
+            }
+
+            private WorkspaceChangeTracker(CodeAnalysisService owner, Workspace workspace, DocumentId documentId, VersionStamp dependentVersion)
+            {
+                _gate = new object();
+
+                _owner = owner;
+                _workspace = workspace;
+                _documentId = documentId;
+
+                _lastVersion = dependentVersion;
+                _resettableDelay = ResettableDelay.CompletedDelay;
+
+                ConnectEvents(subscription: true);
+            }
+
+            private void ConnectEvents(bool subscription)
+            {
+                // this is only place lock is used.
+                // we have a lock here so that subscription and unsubscription of the two events
+                // are happening atomic, but that doesn't mean there is no possiblity of race here
+                // theoradically, there can be a race if connection got disconnected before we subscribe
+                // to OnRpcDisconnected but already in the subscription code path.
+                // but there is no easy way to solve the problem unless Rpc itself provide things like
+                // subscribe only if connection still alive or something
+                lock (_gate)
+                {
+                    if (subscription)
+                    {
+                        _owner.Disconnected += OnDisconnected;
+                        _workspace.WorkspaceChanged += OnWorkspaceChanged;
+                    }
+                    else
+                    {
+                        _owner.Disconnected -= OnDisconnected;
+                        _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+                    }
+                }
+            }
+
+            private void OnDisconnected(object sender, EventArgs e)
+            {
+                ConnectEvents(subscription: false);
+            }
+
+            private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+            {
+                EnqueueUpdate();
+                return;
+
+                void EnqueueUpdate()
```

Comment:
💡 This seems like it would be more appropriate as either private method, or inlined at the single call site.

Reply:
ya. not sure why it got into this shape :)

Reply:
hmm.. it might have been in recursive form and I ended up remove that. anyway, thank you!